### PR TITLE
Replace double ampersand with comma #315

### PR DESCRIPTION
**Pull request by akantsevoi**
Replace double ampersand with comma https://github.com/nicklockwood/SwiftFormat/issues/315

### CODE REVIEW

File: Sources/Options.swift [hunk 1]
```diff
@@ -199,6 +199,7 @@ public struct FormatOptions: CustomStringConvertible {
     public var removeSelf: Bool
     public var experimentalRules: Bool
     public var fragment: Bool
+    public var commasInsteadOfAmpersands: Bool
```
Comment (nicklockwood) [inline Sources/Options.swift]: I don’t think there’s any need to add an option for this. Since it’s possible to disable the rule using `—disable commasInsteadOfAmpsersands`.

In general I only add an option if the rule needs multiple configuration options beyond just enable/disable.

File: Tests/RulesTests.swift [hunk 2]
```diff
@@ -7014,4 +7017,70 @@ class RulesTests: XCTestCase {
...
+        XCTAssertEqual(try format(input + "\n", rules: FormatRules.default), output + "\n")
+    }
+
+    func testGuardAmpersandReplacedWithFunction() {
+        let input = "if functionReturnsBool() && variable\nelse { return }"
+        let output = "if functionReturnsBool(), variable\nelse { return }"
+
+        XCTAssertEqual(try format(input, rules: [FormatRules.commasInsteadOfAmpersands]), output)
+        XCTAssertEqual(try format(input + "\n", rules: FormatRules.default), output + "\n")
+    }
```
Comment (nicklockwood) [inline Tests/RulesTests.swift]: I’m curious how it handles cases like

`if foo || bar && baz {}`

Since replacing && with , in that case would change the operator precedence.

Comment (akantsevoi) [reply Tests/RulesTests.swift]: I'll add tests on that cases:
```
let input = "if foo || bar && baz {}"
let output = "if foo || bar, baz {}"
```

Comment (akantsevoi) [reply Tests/RulesTests.swift]: The precedence doesn't change.
For example those expressions
`if true && true || false`
`if true, true || false`

are equal.

File: Sources/Options.swift (hunk 1 above)
Comment (akantsevoi) [reply Sources/Options.swift]: I see. But how, in that case I can disable an option for that test:
```
func testNestedWrappedIfIndents() {
        let input = "if foo {\nif bar &&\n(baz ||\nquux) {\nfoo()\n}\n}"
        let output = "if foo {\n    if bar &&\n        (baz ||\n            quux) {\n        foo()\n    }\n}"
        let options = FormatOptions(commasInsteadOfAmpersands: false)
        XCTAssertEqual(try format(input, rules: [FormatRules.indent]), output)
        XCTAssertEqual(try format(input + "\n", rules: FormatRules.default, options: options), output + "\n")
    }
```

File: Tests/RulesTests.swift (hunk 2 above)
Comment (akantsevoi) [reply Tests/RulesTests.swift]: Look at https://github.com/nicklockwood/SwiftFormat/pull/316/commits/05de6a5bf7dcd7817e52c87f89ab553279b662e2

File: Sources/Options.swift (hunk 1 above)
Comment (nicklockwood) [reply Sources/Options.swift]: @akantsevoi 
```swift
let rules = FormatRules.all(except: ["commasInsteadOfAmpersands"])
XCTAssertEqual(try format(input, rules: rules, output)
```

Comment (akantsevoi) [reply Sources/Options.swift]: Oh, I see, and in that case I shouldn't check it in a rule?:
```
@objc public class func commasInsteadOfAmpersands(_ formatter: Formatter) {
        guard formatter.options.commasInsteadOfAmpersands else { return }
```

Comment (nicklockwood) [reply Sources/Options.swift]: @akantsevoi no, the rule method will not be run if it's disabled, so no need for the extra check.

Comment (akantsevoi) [reply Sources/Options.swift]: Good, will update PR soon.

File: Tests/RulesTests.swift (hunk 2 above)
Comment (nicklockwood) [reply Tests/RulesTests.swift]: @akantsevoi sorry, I gave a bad example. Here is a better one:

```swift
if false && true || true { // prints true
    print("true")
} else {
    print("false")
}

if false, true || true { // prints false
    print("true")
} else {
    print("false")
}
```

Comment (akantsevoi) [reply Tests/RulesTests.swift]: Hm, I see. Thanks for the example.
I think in that case I should keep it as is.
For example if we have expression in commas (true && false) - I do nothing with this.
What do you think?

Comment (nicklockwood) [reply Tests/RulesTests.swift]: @akantsevoi agreed. Ideally it would take the precedence into account - i.e. for `foo || bar && baz` it would replace the `&&`, but for `foo && bar || baz` it wouldn't, but that may be too complicated to do reliably. I'd be happy with a solution that just ignores mixed operators - I can always enhance it later.

Comment (nicklockwood) [reply Tests/RulesTests.swift]: @akantsevoi here's another nasty case. Perhaps ignoring the && if the clause contains an `=` or `case` would be a good precaution?

```swift
let foo = true
let bar = false

if case let a = foo && bar { // prints "false"
    print(a)
}

if case let a = foo, bar { // prints nothing
    print(a)
}
```

Comment (akantsevoi) [reply Tests/RulesTests.swift]: This feature becomes more and more useless feature))

Comment (nicklockwood) [reply Tests/RulesTests.swift]: @akantsevoi I don't  agree at all! It's something I've wanted to add for a while, and these edge cases are very rare compared to the common case that will work correct - it's just important that a formatter doesn't break code, especially in subtle ways like changing the behavior but still compiling  :-)

Comment (akantsevoi) [reply Tests/RulesTests.swift]: It was a joke) Agree with you

**Comment by nicklockwood** [comment]
@akantsevoi another case we should support is `while foo, bar { ... }`, which works the same as `if` and `guard`.

I'm not sure if there are any other cases where you can use commas in place of `&&`. I don't think it's allowed in `for` loops or `where` clauses.

File: Sources/Rules.swift [hunk 3]
```diff
@@ -3274,6 +3274,41 @@ extension FormatRules {
             }
         }
     }
+
+    /// Replace `&&` to `,` in if or guard closure
+    @objc public class func commasInsteadOfAmpersands(_ formatter: Formatter) {
+        guard formatter.options.commasInsteadOfAmpersands else { return }
+        guard let firstToken = formatter.tokens.first,
```
Comment (nicklockwood) [inline Sources/Rules.swift]: It looks like this rule will only work if the first token in the whole file is an `if` or `guard` keyword. Rules are not automatically applied to every token, they are applied once to the whole file, so I think you need to wrap the body of this entire rule in a ` formatter.forEachToken { ... }`.

File: Sources/Rules.swift [hunk 4]
```diff
@@ -3274,6 +3274,41 @@ extension FormatRules {
...
+            case let Token.keyword(keyword) = firstToken,
+            keyword == "if" || keyword == "guard"
+        else { return }
+
+        var previous = (0, firstToken)
+        var openCommansCounter = 0
+        formatter.forEachToken { index, token in
```
Comment (nicklockwood) [inline Sources/Rules.swift]: You can't nest `formatter.forEachToken { }` calls, so you'll need to rethink the logic for this once you've dealt with the above comment. I suggest use the `formatter.next/last` methods to navigate through the tokens until you reach the end of the `if`/`guard` statement.

File: Sources/Rules.swift (hunk 3 above)
Comment (nicklockwood) [reply Sources/Rules.swift]: There are smarter variants of `formatter.forEachToken`, so you probably want to use something like:

```swift
formatter.forEachToken(where: { $0 == .keyword("if") || $0 == .keyword("guard") || $0 == .keyword("where") }) {
    ... your rule logic
}
```

File: Tests/RulesTests.swift (hunk 2 above)
Comment (akantsevoi) [reply Tests/RulesTests.swift]: One more case:
```
func someFuncWithClosure(closure: (Int) -> Bool) -> Bool {
    return closure(4)
}

if true && someFuncWithClosure{$0 == 4} {
    print("true")
}
```

Comment (akantsevoi) [reply Tests/RulesTests.swift]: Do you have any idea how to resolve the case above? Because for finish replacing I check `Token.startOfScope("{")` But in that case I'm not able to do it.

Comment (nicklockwood) [reply Tests/RulesTests.swift]: @akantsevoi SwiftFormat can't disambiguate that, but that's OK because it's not valid code - you aren't allowed to use trailing closure syntax inside an if condition because the Swift compiler can't disambiguate it either 😄.

It would have to be written as:

```swift
if true && someFuncWithClosure({$0 == 4}) {
    print("true")
}
```

So you can handle that by just ignoring anything inside `( ... )` (which I think you do anyway, right?).

Comment (nicklockwood) [reply Tests/RulesTests.swift]: @akantsevoi FYI, if you use something like `let endIndex = formatter.index(of: .startOfScope("{"), after: startIndex)` it will automatically skip over anything inside `( ... )` because it only looks for tokens at the current scope level. That should make it a bit easier to write your rule.

Comment (akantsevoi) [reply Tests/RulesTests.swift]: Unfortunately this code works in my environment:
```
func someFuncWithClosure(closure: (Int) -> Bool) -> Bool {
    return closure(4)
}

if true && someFuncWithClosure{$0 == 4} {
    print("true")
}
```

Comment (nicklockwood) [reply Tests/RulesTests.swift]: @akantsevoi oh crap... they must have added support for that in Xcode 10 because it didn't work before 🤔. I guess we can disambiguate it by checking if the first  `{ ... }` is immediately followed by another  `{`, but that means I probably need to revisit a bunch of other rules too 😬

Comment (nicklockwood) [reply Tests/RulesTests.swift]: @akantsevoi investigating further, this is allowed:

`if true && someFuncWithClosure {$0 == 4}, true {`

But this isn't:

`if true && someFuncWithClosure {$0 == 4} && true {`

So the rule needs to be something like (this is untested and could probably be nicer):

```swift
// find endIndex
guard var endIndex = formatter.index(of: .startOfScope("{"), after: startIndex) else {
    return // malformed code
}
// test if this is really the end
while let endOfScopeIndex = formatter.endOfScope(at: endIndex),
    let nextIndex = formatter.index(of: .nonSpaceOrCommentOrLinebreak, after: endOfScopeIndex) {
    let token = formatter.tokens[nextIndex]
    if token == .delimiter(",") {
        guard var newEndIndex = formatter.index(of: .startOfScope("{"), after: startIndex) else {
            return // malformed code
        }
        endIndex = newEndIndex
    } else if token == .startOfScope("{") {
        endIndex = newEndIndex
    } else {
        break // found it
    }
}
```

Comment (akantsevoi) [reply Tests/RulesTests.swift]: @nicklockwood I've got an idea.

Comment (nicklockwood) [reply Tests/RulesTests.swift]: @akantsevoi looking into it some more, I'm pretty sure that it's a bug that this compiles at all, and it will be so rare that it's unlikely anyone will encounter it.

In any case, since the rule doesn't care about the if statement body, the only consequence of mistaking the first { for the start of the body will be that it ignores any &&s that may appear later in the same if condition, which is not a major problem and won't cause any code corruption.

So I think it's fine to simply not worry about this case.

Comment (akantsevoi) [reply Tests/RulesTests.swift]: Yes, but the problem is how to separate function call, variable and function call with closure. Because first two variants - absolutely correct and legible for replacing.

Comment (nicklockwood) [reply Tests/RulesTests.swift]: Ah, I see.

Comment (akantsevoi) [reply Tests/RulesTests.swift]: @nicklockwood unfortunately I don't have any idea how to resolve this issue in the current situation. Maybe we can improve Token type and add some more information into it?

Comment (nicklockwood) [reply Tests/RulesTests.swift]: @akantsevoi understood. I'll take a look at your implementation and see if I can find a workaround. Thanks for all the work you've put in so far!

**Comment by nicklockwood** [comment]
@akantsevoi this feature is now implemented in SwiftFormat 0.36.0. Thanks for the PR!

**Comment by akantsevoi** [comment]
@nicklockwood  it was pleasure. Thanks a lot for such a great tool
